service/policy/db/key_access_server_registry.go

```diff
@@ -830,3 +828,188 @@ func (c PolicyDBClient) rotatePublicKeyTables(ctx context.Context, oldKeyID, new
+		defaultKeys, err = db.DefaultKasKeysProtoJSON(keys)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return defaultKeys, nil
+}
+
+func (c PolicyDBClient) GetDefaultKasKeyByMode(ctx context.Context, tdfType kasregistry.TdfType) (*kasregistry.DefaultKasKey, error) {
+	key, err := c.getDefaultKasKeyByMode(ctx, pgtypeText(tdfType.String()))
+	if err != nil && !errors.Is(db.WrapIfKnownInvalidQueryErr(err), db.ErrNotFound) {
+		c.logger.Error("GetDefaultKasKeyByMode", "error", err)
+		return nil, db.WrapIfKnownInvalidQueryErr(err)
+	}
+
+	var defaultKey *kasregistry.DefaultKasKey
+	if len(key) > 0 {
+		defaultKey = &kasregistry.DefaultKasKey{}
+		err = db.UnmarshalDefaultKasKey(key, defaultKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return defaultKey, nil
+}
+
+func isAlgValidForNano(alg policy.Algorithm) bool {
+	switch alg {
+	case policy.Algorithm_ALGORITHM_EC_P256, policy.Algorithm_ALGORITHM_EC_P384, policy.Algorithm_ALGORITHM_EC_P521:
+		return true
+	case policy.Algorithm_ALGORITHM_RSA_2048, policy.Algorithm_ALGORITHM_RSA_4096, policy.Algorithm_ALGORITHM_UNSPECIFIED:
+		return false
+	default:
+		return false
+	}
+}
+
+func (c PolicyDBClient) SetDefaultKey(ctx context.Context, r *kasregistry.SetDefaultKeyRequest) (*kasregistry.SetDefaultKeyResponse, error) {
+	var identifier any
+	switch r.GetActiveKey().(type) {
+	case *kasregistry.SetDefaultKeyRequest_Id:
+		identifier = &kasregistry.GetKeyRequest_Id{
+			Id: r.GetId(),
+		}
+	case *kasregistry.SetDefaultKeyRequest_Key:
+		identifier = &kasregistry.GetKeyRequest_Key{
+			Key: r.GetKey(),
+		}
+	}
+	keyToSet, err := c.GetKey(ctx, identifier)
+	if err != nil {
+		return nil, err
+	}
+
+	previousDefaultKey, err := c.GetDefaultKasKeyByMode(ctx, r.GetTdfType())
+	if err != nil {
+		return nil, err
+	}
+
+	// If default key is nano, cipher must be ECC.
+	if r.GetTdfType() == kasregistry.TdfType_TDF_TYPE_NANO && !isAlgValidForNano(keyToSet.GetKey().GetKeyAlgorithm()) {
+		return nil, fmt.Errorf("key algorithm %s is not valid for TDF type NANO", keyToSet.GetKey().GetKeyAlgorithm().String())
+	}
+
+	// A trigger is set for BEFORE INSERT which will update the
+	// the key reference to the one being inserted, if present.
+	// If not, the insert will continue.
+	_, err = c.Queries.setDefaultKasKey(ctx, setDefaultKasKeyParams{
+		KeyAccessServerKeyID: pgtypeUUID(keyToSet.GetKey().GetId()),
+		TdfType:              r.GetTdfType().String(),
+	})
+	if err != nil {
+		return nil, db.WrapIfKnownInvalidQueryErr(err)
+	}
+
+	// Get the new default key.
+	newDefaultKey, err := c.GetDefaultKasKeyByMode(ctx, r.GetTdfType())
+	if err != nil {
+		return nil, err
+	}
+
+	// Set wellknown config
+	if err := c.SetWellKnownConfig(ctx); err != nil {
+		return nil, err
+	}
+
+	return &kasregistry.SetDefaultKeyResponse{
+		NewDefaultKasKey:      newDefaultKey,
+		PreviousDefaultKasKey: previousDefaultKey,
+	}, nil
+}
+
+func (c PolicyDBClient) SetWellKnownConfig(ctx context.Context) error {
```

Comment:
We might want to rename this because I could see other areas maybe in policy we want to update the well-known.